packages/tools/playground/src/components/monacoComponent.tsx

```diff
@@ -1,68 +1,821 @@
+        const set = new Set(files);
+        const kept = order.filter((p: string) => set.has(p));
+        const extras = files.filter((p: string) => !kept.includes(p));
+        return kept.concat(extras);
+    }
+
+    private _orderedFiles(): string[] {
+        const { files, order } = this.state;
+        if (!order.length) {
+            return files;
+        }
+        const inSet = new Set(files);
+        const merged = order.filter((p) => inSet.has(p));
+        for (const f of files) {
+            if (!merged.includes(f)) {
+                merged.push(f);
+            }
+        }
+        return merged;
+    }
+
+    private _scrollActiveIntoView() {
+        const el = document.querySelector<HTMLDivElement>(`.pg-tab[data-path="${CSS.escape(this.state.active)}"]`);
+        el?.scrollIntoView({ inline: "nearest", block: "nearest", behavior: "smooth" });
+    }
+
+    private _defaultNewFileName(): string {
+        const isTS = this.props.globalState.language !== "JS";
+        const base = isTS ? "new.ts" : "new.js";
+        let idx = 0;
+        let candidate = base;
+        while ((this.props.globalState.files as any)[this._toInternal(candidate)]) {
+            idx++;
+            const parts = base.split(".");
+            const ext = parts.pop();
+            candidate = `${parts.join(".")}${idx}.${ext}`;
+        }
+        return candidate;
+    }
+
+    private _defaultDuplicateName(path: string): string {
+        const baseFull = path.replace(/\.(\w+)$/, "");
+        const ext = (path.match(/\.(\w+)$/) || ["", ""])[1];
+        let i = 1;
+        let candidate = `${baseFull}.copy${i}${ext ? "." + ext : ""}`;
+        while ((this.props.globalState.files as any)[candidate]) {
+            i++;
+            candidate = `${baseFull}.copy${i}${ext ? "." + ext : ""}`;
+        }
+        return this._toDisplay(candidate);
+    }
+
+    private _openLocalSessionDialog = () => {
+        this.setState({ sessionDialogOpen: true });
+    };
+
+    private _closeLocalSessionDialog = () => {
+        this.setState({ sessionDialogOpen: false });
+    };
+
+    private _openDialog(type: DialogKind, title: string, initialValue: string, targetPath?: string) {
+        this._closeCtxMenu();
+        this.setState({
+            fileDialog: { open: true, type, title, initialValue, targetPath },
+        });
+    }
+
+    private _closeFileDialog = () => {
+        this.setState({ fileDialog: { open: false, type: null, title: "", initialValue: "" } });
+    };
+
+    private _confirmFileDialog = (filename: string) => {
+        const { type, targetPath } = this.state.fileDialog;
+        if (!type) {
+            return;
+        }
+
+        const internal = this._toInternal(filename.trim());
+        const exists = (this.props.globalState.files as any)[internal];
+
+        if (type === "create") {
+            if (exists) {
+                alert("A file with that name already exists.");

```

Comment:
If I'm reading this right, it's trying to prevent adding a new file with the same name as an existing file. But when I test this scenario, I can add a file with the same name (e.g. new.ts) and then the tabs wonk out.

Reply:
I found the bug, was truthiness when it should've been checking for existence, so when a new file existed but was empty was giving a false negative. Fixed this